presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/GroupInnerJoinsByConnector.java

```diff
@@ -0,0 +1,663 @@
+import static com.facebook.presto.spi.connector.ConnectorCapabilities.SUPPORTS_JOIN_PUSHDOWN;
+import static com.facebook.presto.spi.plan.JoinType.INNER;
+import static com.facebook.presto.sql.planner.VariablesExtractor.extractUnique;
+import static com.facebook.presto.sql.planner.optimizations.JoinNodeUtils.toRowExpression;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.filter;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * GroupInnerJoinsByConnector Optimizer:
+ * This optimizer attempts to group TableScanNode's of an inner-join graph that belong to the same connector
+ * This allows those connectors that can participate in join pushdown, to rewrite these sources to a new TableScanNode that represents the result of the pushed down join
+ *
+ * Example:
+ * Before Transformation:
+ *   --OutputNode
+ *      `-- InnerJoin1
+ *          |-- InnerJoin2 (Left)
+ *          |   |-- TableScanNode1 (Left)
+ *          |   |   `-- TableHandle1
+ *          |   `-- TableScanNode2 (Right)
+ *          |       `-- TableHandle2
+ *          `-- TableScanNode3 (Right)
+ *              `-- TableHandle3
+ *
+ * Suppose that TableScanNode1, TableScanNode2 and TableScanNode3 are from the same catalog.
+ *
+ * After Transformation:
+ *   --OutputNode
+ *   `-- TableScanNode (with all the details of the three TableScanNodes)
+ *       `-- Set<ConnectorTableHandle> (ConnectorHandleSet)
+ *          `-- TableHandle1
+ *          `-- TableHandle2
+ *          `-- TableHandle3
+ */
+public class GroupInnerJoinsByConnector
+        implements PlanOptimizer
+{
+    private static final Logger logger = Logger.get(GroupInnerJoinsByConnector.class);
+    private final FunctionResolution functionResolution;
+    private final DeterminismEvaluator determinismEvaluator;
+    private final Metadata metadata;
+    private boolean isEnabledForTesting;
+
+    public GroupInnerJoinsByConnector(Metadata metadata)
+    {
+        this.functionResolution = new FunctionResolution(metadata.getFunctionAndTypeManager().getFunctionAndTypeResolver());
+        this.determinismEvaluator = new RowExpressionDeterminismEvaluator(metadata.getFunctionAndTypeManager());
+        this.metadata = metadata;
+    }
+
+    @Override
+    public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (isEnabled(session)) {
+            PlanNode rewrittenPlan = SimplePlanRewriter.rewriteWith(new Rewriter(functionResolution, determinismEvaluator, idAllocator, metadata, session), plan);
+            return PlanOptimizerResult.optimizerResult(rewrittenPlan, !rewrittenPlan.equals(plan));
```

Comment:
We've been preferring the use of the IterativeOptimizer over PlanOptimizer.  

Since your rewrite can very explicitly determine if the plan is changed (by using `rewrittenPlan.equals(plan)`) modifying this class to instead implement an iterative optimizer rule (i.e `implements Rule<JoinNode>`) should be pretty straightforward.

This should reduce the test burden as well - you will be able to use a `RuleTester` and its associated helpers directly

Reply:
Hi @aaneja , @Haritha-Koloth 

Below are the reasons to use PlanOptimizer instead of Iterative optimizer Rule.

1) Need to visit Multiple nodes (visitJoin, visitFilter, etc.). Using a Rule we may able to perform operations on a single node only, I guess.

2) Repeated application of JoinNode Rule, due to child join.

Not sure these cases can resolved through a Single Iterative Optimizer Rule. For handling filter node, required a Rule<FilterNode> and for join required another Rule<JoinNode>.

Reply:
@aaneja Thanks! We are exploring the potential and viability of the above suggestions.

Reply:
Hi @Haritha-Koloth and @aaneja 

I think we already did some good test coverage for this using PlanOptimizer itself.

Do we have any additional benefit for changing to  optimizer Rule? If we use optimizer Rule how we could handlw non-equijoin scenario where we have to visit FilterNode and JoinNode

Reply:
@Ajas-Mangal As per the discussion with @aaneja, the usage of IterativeOptimizer is being preferred over PlanOptimizer implementations and there is a potential advantage of opening up more nodes for pushdown. 

As for the non-equijoin scenario, I guess we can have multiple rule matchers to handle the filter and join nodes.

I am trying out the suggested changes, will update here if met with some blockers/concerns.